promql/engine.go

```diff
@@ -606,88 +579,73 @@ func (ev *evaluator) recover(errp *error) {
 	if e == nil {
 		return
 	}
-	if _, ok := e.(runtime.Error); ok {
+	if err, ok := e.(runtime.Error); ok {
 		// Print the stack trace but do not inhibit the running application.
 		buf := make([]byte, 64<<10)
 		buf = buf[:runtime.Stack(buf, false)]
 
 		level.Error(ev.logger).Log("msg", "runtime panic in parser", "err", e, "stacktrace", string(buf))
-		*errp = fmt.Errorf("unexpected error")
+		*errp = fmt.Errorf("unexpected error: %s", err)
 	} else {
 		*errp = e.(error)
 	}
 }
 
-// evalScalar attempts to evaluate e to a Scalar value and errors otherwise.
-func (ev *evaluator) evalScalar(e Expr) Scalar {
-	val := ev.eval(e)
-	sv, ok := val.(Scalar)
-	if !ok {
-		ev.errorf("expected Scalar but got %s", documentedType(val.Type()))
-	}
-	return sv
+func (ev *evaluator) Eval(expr Expr) (v Value, err error) {
+	defer ev.recover(&err)
+	return ev.eval(expr), nil
 }
 
-// evalVector attempts to evaluate e to a Vector value and errors otherwise.
-func (ev *evaluator) evalVector(e Expr) Vector {
-	val := ev.eval(e)
-	vec, ok := val.(Vector)
-	if !ok {
-		ev.errorf("expected instant Vector but got %s", documentedType(val.Type()))
-	}
-	return vec
-}
+// Extra information and caches for evaluating a single node across steps.
+type evalNodeHelper struct {
+	// Evaluation timestamp.
+	ts int64
+	// Vector that can be used for output.
+	out Vector
 
-// evalInt attempts to evaluate e into an integer and errors otherwise.
-func (ev *evaluator) evalInt(e Expr) int64 {
-	sc := ev.evalScalar(e)
-	if !convertibleToInt64(sc.V) {
-		ev.errorf("Scalar value %v overflows int64", sc.V)
-	}
-	return int64(sc.V)
-}
+	// Caches.
+	dmn                          map[uint64]labels.Labels      // dropMetricName and label_*.
+	sigf                         map[uint64]uint64             // signatureFunc.
+	signatureToMetricWithBuckets map[uint64]*metricWithBuckets // funcHistogramQuantile.
+	regex                        *regexp.Regexp                // label_replace.
```

Comment:
I'm not sure it makes sense to document that here, the functions they're used in are mentioned so you can see what they're for.

Reply:
Might be useful to move the comments onto separate rows preceding the field (possibly prefixed with "Used by") so that it works as documentation (e.g. to be shown as a tooltip by an IDE).

Reply:
Done